File: tests/integration-tests/csharp/TaskExtensions.cs
namespace csharp;

public static class TaskExtensions
{
    public static async Task OrTimeout(this Task task, int millisecondsDelay = 5000)
    {
        var completed = await Task.WhenAny(task, Task.Delay(millisecondsDelay));
        if (!task.IsCompleted)
        {
            throw new TimeoutException();
        }
    }
}

File: tests/integration-tests/java/TaskExtensions.java
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class TaskExtensions {
    public static <T> CompletableFuture<T> orTimeout(CompletableFuture<T> future, int millisecondsDelay) {
        CompletableFuture<T> timeoutFuture = new CompletableFuture<>();
        CompletableFuture.delayedExecutor(millisecondsDelay, TimeUnit.MILLISECONDS).execute(() -> timeoutFuture.completeExceptionally(new TimeoutException()));
        return future.applyToEither(timeoutFuture, result -> result);
    }
}
